# Exercise 4.8

## 1.
### Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe

## Importing Libraries

In [1]:
# Importing libraries

import os
import numpy as np
import pandas as pd

## Importing Data

In [2]:
# Set path

path = r"C:\Users\abiha\OneDrive\Documents\IC Project Folder"

In [3]:
# Importing dataset

ords_prods_merge = pd.read_pickle(r"C:\Users\abiha\OneDrive\Documents\IC Project Folder\2. Data\Prepared Data\order_products_merged_newest.pkl")

## 2.
### In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [4]:
# Finding the average of the order_number column, grouped by department

ords_prods_merge.groupby('department_id').agg({'order_number' : ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## 3.
### Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

#### Compared with the results of the subset, the mean values vary quite significantly across almost every department. 
#### Also, the subset does not return any results for the highest performing departments, 10(bulk) 18(babies). nb department id 21 (missing) not included as I am assuming that could be multiple departments that have not been entered into dataframe
#### The subset would also suggest that the difference between average orders of departments is higher. Showing some averages as low as 11.2 when the reality is that most fall somewhere between 15.2 and 17.8

## 4.

### Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [5]:
# Create 'max_order' column

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [6]:
# Create flag to show loyal customers

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [7]:
# Create flag to show reglular customers

ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [8]:
# Create flag to show new customers

ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [9]:
# Count frequency of each 'loyalty flag'

ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

## 5.
### The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [10]:
# Check basic stats of product prices, grouped by loyalty category

ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean','min','max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

## 6.
### The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
#### If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
#### If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [11]:
# Create average order column in df

ords_prods_merge['average_order'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [12]:
# Check column visible

ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   

    days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                      NaN            1         196                  1   
1                     15.0            0         196                  1   
2                     21.0            0         196                  1   
3                     29.0            0         196                  1   
4                     28.0            0         196                  1   
5                     19.0            0         196                  1   
6                     20.0            0         196                  1   
7                     14.0            0         196                  2   
8                      0.0            0         196                  4   
9                     30.0            0         196                  1   
10                     7.0            0         196                  2   
11                     8.0            0         196                  6   
12                     7.0            0         196                  1   
13                     7.0            0         196                  1   
14                    14.0            0         196                  1   

    reordered  ... aisle_id  department_id prices    price_range_loc  \
0           0  ...       77              7    9.0  Mid-range product   
1           1  ...       77              7    9.0  Mid-range product   
2           1  ...       77              7    9.0  Mid-range product   
3           1  ...       77              7    9.0  Mid-range product   
4           1  ...       77              7    9.0  Mid-range product   
5           1  ...       77              7    9.0  Mid-range product   
6           1  ...       77              7    9.0  Mid-range product   
7           1  ...       77              7    9.0  Mid-range product   
8           1  ...       77              7    9.0  Mid-range product   
9           1  ...       77              7    9.0  Mid-range product   
10          0  ...       77              7    9.0  Mid-range product   
11          1  ...       77              7    9.0  Mid-range product   
12          1  ...       77              7    9.0  Mid-range product   
13          1  ...       77              7    9.0  Mid-range product   
14          1  ...       77              7    9.0  Mid-range product   

       busiest_day     Busiest_days busiest_period_of_day max_order  \
0   Regularly busy   Regularly busy        Average orders        10   
1   Regularly busy  Least busy days        Average orders        10   
2   Regularly busy  Least busy days           Most orders        10   
3       Least busy  Least busy days        Average orders        10   
4       Least busy  Least busy days           Most orders     

In [13]:
# Create flag for low spenders

ords_prods_merge.loc[ords_prods_merge['average_order'] < 10, 'spender_flag'] = 'Low spender'

In [14]:
# Create flag for high spenders

ords_prods_merge.loc[ords_prods_merge['average_order'] >= 10, 'spender_flag'] = 'High spender'

In [15]:
# Frequency count for spender category

ords_prods_merge['spender_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: spender_flag, dtype: int64

## 7.

### In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
#### If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
#### If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
#### If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [16]:
# Create column to show median of days_since_prior_order per user

ords_prods_merge['median_order_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [17]:
# Checking new column

ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   

    days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                      NaN            1         196                  1   
1                     15.0            0         196                  1   
2                     21.0            0         196                  1   
3                     29.0            0         196                  1   
4                     28.0            0         196                  1   
5                     19.0            0         196                  1   
6                     20.0            0         196                  1   
7                     14.0            0         196                  2   
8                      0.0            0         196                  4   
9                     30.0            0         196                  1   
10                     7.0            0         196                  2   
11                     8.0            0         196                  6   
12                     7.0            0         196                  1   
13                     7.0            0         196                  1   
14                    14.0            0         196                  1   

    reordered  ... prices    price_range_loc     busiest_day     Busiest_days  \
0           0  ...    9.0  Mid-range product  Regularly busy   Regularly busy   
1           1  ...    9.0  Mid-range product  Regularly busy  Least busy days   
2           1  ...    9.0  Mid-range product  Regularly busy  Least busy days   
3           1  ...    9.0  Mid-range product      Least busy  Least busy days   
4           1  ...    9.0  Mid-range product      Least busy  Least busy days   
5           1  ...    9.0  Mid-range product  Regularly busy   Regularly busy   
6           1  ...    9.0  Mid-range product  Regularly busy     Busiest days   
7           1  ...    9.0  Mid-range product  Regularly busy     Busiest days   
8           1  ...    9.0  Mid-range product  Regularly busy     Busiest days   
9           1  ...    9.0  Mid-range product      Least busy  Least busy days   
10          0  ...    9.0  Mid-range product  Regularly busy     Busiest days   
11          1  ...    9.0  Mid-range product  Regularly busy   Regularly busy   
12          1  ...    9.0  Mid-range product  Regularly busy   Regularly busy   
13          1  ...    9.0  Mid-range product  Regularly busy     Busiest days   
14          1  ...    9.0  Mid-range product  Regularly busy     Busiest days   

    busiest_period_of_day  max_order      loyalty_flag average_order  \
0          Average orders         10      New customer      6.367797   
1          Average orders         10      New customer      6.367797   
2             Most orders         10      New customer   

In [18]:
# Create flag for non-frequent customers

ords_prods_merge.loc[ords_prods_merge['median_order_frequency'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [19]:
# Create flag for regular customers

ords_prods_merge.loc[(ords_prods_merge['median_order_frequency'] > 10) & (ords_prods_merge['median_order_frequency'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [20]:
# Create flag for frequent customers

ords_prods_merge.loc[ords_prods_merge['median_order_frequency'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [21]:
# Check frequency count for customer frequency categories

ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_frequency_flag, dtype: int64

### Exporting dataframe as pickle file in 'Prepared Data' folder

In [ ]:
ords_prods_merge.to_pickle(os.path.join(path, '2. Data', 'Prepared Data', 'Orders_products_merged_4_8.pkl'))